lib/OpenQA/WebAPI/Controller/Admin/Needle.pm

```diff
@@ -77,18 +77,46 @@ sub ajax {
             my ($needles) = @_;
             my @data;
             my %modules;
+            my %needle_dir_real_paths;
 
-            my %needle_dir_map = map { $_->path => $_ } $needles->all;
-
+            my $schema = $self->schema;
             while (my $n = $needles->next) {
-                my $hash = {
+                my $filename = $n->filename;
+                my $hash     = {
                     id        => $n->id,
                     directory => $n->directory->name,
-                    filename  => $n->filename,
+                    filename  => $filename,
                 };
+                my $needle_dir_path = $n->directory->path;
+                my $real_needle_dir_id;
+
+                if ($needle_dir_real_paths{$needle_dir_path}) {
```

Comment:
I'd prefer an `my $needle_dir_real_path = $needle_dir_real_paths{$needle_dir_path}` within the if-condition and use that in the block instead of repeating `$needle_dir_real_paths{$needle_dir_path}` two more times.